core/relayer/deliver.go

```diff
@@ -0,0 +1,201 @@
+package relayer
+
+import (
+	//"context"
+	//"encoding/hex"
+	//"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc"
+	//"github.com/golang/protobuf/proto"
+
+	//"github.com/xuperchain/xuperchain/core/contract"
+	//"github.com/xuperchain/xuperchain/core/global"
+	"github.com/xuperchain/xuperchain/core/pb"
+	//"github.com/xuperchain/xuperchain/core/utxo"
+	//"github.com/xuperchain/xuperchain/core/utxo/txhash"
+	"github.com/xuperchain/xuperchain/core/common"
+	relayerpb "github.com/xuperchain/xuperchain/core/relayer/pb"
+)
+
+type QueryMetaRegister interface {
+	GetLastQueryBlockHeight() int64
+}
+
+type DeliverBlockCommand struct {
+	client            pb.XchainClient
+	Cfg               ChainConfig
+	meta              *relayerpb.DeliverMeta
+	AnchorBlockHeight int64
+	Storage           *Storage
+	QueryMeta         QueryMetaRegister
+}
+
+func (cmd *DeliverBlockCommand) InitXchainClient() error {
+	conn, err := grpc.Dial(cmd.Cfg.RPCAddr, grpc.WithInsecure(), grpc.WithMaxMsgSize(64<<20-1))
+	if err != nil {
+		return err
+	}
+	cmd.client = pb.NewXchainClient(conn)
+	return nil
+}
+
+func (cmd *DeliverBlockCommand) Deliver() error {
+	// 是否有必要调用DeliverAnchorBlockHeader()
+	meta, err := cmd.LoadDeliverMeta()
+	if common.NormalizedKVError(err) == common.ErrKVNotFound {
+		cmd.DeliverAnchorBlockHeader()
+	} else if err != nil {
+		panic("get deliver meta failed, err:" + err.Error())
+	} else {
+		cmd.meta = &relayerpb.DeliverMeta{
+			LastDeliverBlockHeight: meta.GetLastDeliverBlockHeight(),
+		}
+	}
+	cmd.DeliverBlockHeader()
+	return nil
+}
+
+func (cmd *DeliverBlockCommand) DeliverAnchorBlockHeader() error {
+	for {
+		blockBuf, err := cmd.Storage.GetBlockHeaderByHeight(cmd.AnchorBlockHeight)
+		// 本地还没有存储AnchorBlock, 等待
+		if common.NormalizedKVError(err) == common.ErrKVNotFound {
+			time.Sleep(time.Duration(1) * time.Second)
+			continue
+		} else if err != nil {
+			panic("get anchor block header failed, err:" + err.Error())
+		}
+
+		args := make(map[string][]byte)
+		args["blockHeader"] = blockBuf
+		// set preExe parameter
+		moduleName := cmd.Cfg.ContractConfig.ModuleName
+		contractName := cmd.Cfg.ContractConfig.ContractName
+		methodName := cmd.Cfg.ContractConfig.AnchorMethod
+		tx, err := cmd.PreExe(moduleName, contractName, methodName, args)
+		// 如果之前已经调用过initAnchorBlockHeader，则跳过
+		if err != nil && strings.Contains(err.Error(), "only once") {
```

Comment:
contract return status 500 may not be a PRC error, so it's better to check contract response status code?